### PR TITLE
Add distribute-to-one edge type

### PR DESCRIPTION
**Pull request by viliam-durina**
State: feature implemented, but untested

### CODE REVIEW

**Comment by hz-highfive** [comment]
The job `jet-oss-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/jet-oss-pr-builder/3030/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/jet-oss-pr-builder/3030/jet-oss-pr-builder-3030.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>

</details>
</pre>

**Comment by viliam-durina** [comment]
I ended up implementing it only for partitioned edges. First I tried to implement it by not creating queues that aren't used, but this created orphaned processors that didn't receive the DONE_ITEM and therefore the job didn't complete. To fix this I would have to avoid even creating those processors, but those processors could have other input edges with a different routing policy. They can also act as source processors: even though they don't receive any data, they can produce some.

I chose a simpler approach: all the processors, sender&receiver tasklets and queues are created as for a normal distributed edge, but partitions are distributed in a way that the target member gets all of them and the other members get none.

If the target member member is not a cluster member when the execution starts, the job will fail. This is acceptable because the target member has some specific role and must be present.

This edge type will be used in SQL to deliver the results to the initiator and to the client. There will be no fault tolerance for queries that return results to the caller.

**Comment by viliam-durina** [comment]
verify

**Comment by hz-highfive** [comment]
The job `jet-oss-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/jet-oss-pr-builder/3066/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/jet-oss-pr-builder/3066/jet-oss-pr-builder-3066.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   RoutingPolicyDistributedTest.when_distributedToOne_partitioned:80 ? Completion
[INFO] 
[ERROR] Tests run: 2223, Failures: 0, Errors: 1, Skipped: 12
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/DAG.java [hunk 1]
```diff
@@ -363,7 +361,7 @@ public JsonObject toJson(int defaultLocalParallelism) {
             edge.add("to", e.getDestName());
             edge.add("toOrdinal", e.getDestOrdinal());
             edge.add("priority", e.getPriority());
-            edge.add("distributed", e.isDistributed());
+            edge.add("distributedTo", e.getDistributedTo() + "");
```
Comment (mtopolnik) [inline hazelcast-jet-core/src/main/java/com/hazelcast/jet/core/DAG.java:364]: `String.valueOf(e.getDistributedTo())` is nicer than this "append empty string" idiom.

File: hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/ConveyorCollector.java [hunk 2]
```diff
@@ -19,16 +19,19 @@
...
 public class ConveyorCollector implements OutboundCollector {
 
     private final ConcurrentConveyor<Object> conveyor;
     private final int queueIndex;
     private final int[] partitions;
 
-    public ConveyorCollector(ConcurrentConveyor<Object> conveyor, int queueIndex, @Nullable int[] partitions) {
-        this.conveyor = conveyor;
+    public ConveyorCollector(@Nonnull ConcurrentConveyor<Object> conveyor, int queueIndex, @Nullable int[] partitions) {
+        this.conveyor = requireNonNull(conveyor);
```
[the remark this answers is not in the document]
Comment (viliam-durina) [reply hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/ConveyorCollector.java:34]: The `@Nonnull` is not a check, just an annotation which can be ignored.

**Comment by cangencer** [comment]
Do we need to extend SinkImpl API as well so it can be used also from Pipeline API? I also need this feature for QuestDB sink, because the writer needs to be local to the machine QuestDB is running on.

**Comment by viliam-durina** [comment]
@cangencer what about doing it in a next PR? You can create an issue and review this PR independently.

**Pull request merged**